crates/common/src/backup/manager.rs

```diff
@@ -0,0 +1,371 @@
+    /// Creates a new BackupManager instance.
+    ///
+    /// # Arguments
+    /// * `node_kind` - The citrea node kind associated with the BackupManager
+    /// * `base_path` - Optional base_path which will be used for creating backups.
+    /// * `config` - Optional config to override required/optional directories
+    pub fn new(
+        // Todo Wait on https://github.com/chainwayxyz/citrea/pull/1714 and RollupClient enum
+        node_kind: &'static str,
+        base_path: Option<PathBuf>,
+        config: Option<BackupConfig>,
+    ) -> Self {
+        Self {
+            node_kind,
+            base_path,
+            databases: HashMap::new(),
+            l1_processing_lock: Mutex::new(()),
+            l2_processing_lock: Mutex::new(()),
+            config: config.unwrap_or_default(),
+        }
+    }
+
+    /// Acquires a lock for L1 block processing.
+    /// Should be held while any L1 block operation is taken
+    pub async fn start_l1_processing(&self) -> MutexGuard<'_, ()> {
+        self.l1_processing_lock.lock().await
+    }
+
+    /// Acquires a lock for L2 block processing.
+    /// Should be held while any L2 block operation is taken
+    pub async fn start_l2_processing(&self) -> MutexGuard<'_, ()> {
+        self.l2_processing_lock.lock().await
+    }
+
+    /// Register a database with the backup manager.
+    ///
+    /// # Arguments
+    /// * `path` - Identifier/path for the database (e.g. "ledger", "state", "native-db")
+    /// * `db` - Reference to the database instance
+    ///
+    /// # Returns
+    /// * `Ok(())` if registration succeeds
+    /// * `Err` if the path is not in the configured backup directories
+    ///
+    /// # Errors
+    /// Returns error if the provided path is not present in backup_dirs configuration
+    pub fn register_database(
+        &mut self,
+        path: String,
+        db: Arc<sov_schema_db::DB>,
+    ) -> anyhow::Result<()> {
+        ensure!(
+            self.config.backup_dirs.contains(&path),
+            "Unexpect database identifier"
+        );
+        self.databases.insert(path, db);
+        Ok(())
+    }
+
+    /// Creates a backup of all the databases held in config at the specified path.
+    ///
+    /// Acquires both L1 and L2 processing locks to ensure consistency between dbs
+    ///
+    /// # Arguments
+    /// * `path`      - Base directory where the backup will be created
+    /// * `l2_height` - L2 height at which the backup was created
+    ///
+    /// # Returns
+    /// Information about the created backup including block height, path and timestamp
+    pub(super) async fn create_backup(
+        &self,
+        path: Option<PathBuf>,
+        l2_height: u64,
+    ) -> anyhow::Result<CreateBackupInfo> {
+        let backup_path = path
+            .as_ref()
+            .or(self.base_path.as_ref())
+            .context("Missing path and no backup_path found in config.")?;
+
+        let l1_lock = self.l1_processing_lock.lock().await;
```

Comment:
What are these locks for? If we need a FS lock, we can use https://docs.rs/file-lock/latest/file_lock/

Reply:
Ok, understood. Do we produce l1 blocks in parallel while producing l2 blocks or we are blocked by l1 production? If we are blocked, then there is no need for 2 mutexes?